### PR TITLE
Major rewrite of aws-simple

### PR DESCRIPTION
**Pull request by clebert**
- [x] Rewrite of the CDK bindings using the new configuration format
  - [x] `aws-simple synthesize [options]`
- [x] Rewrite of the SDK bindings using the new configuration format
  - [x] `aws-simple upload [options]`
  - [x] `aws-simple list [options]`
  - [x] `aws-simple delete [options]`
  - [x] `aws-simple tag [options]`
  - [x] `aws-simple purge [options]`
  - [x] `aws-simple flush-cache [options]`
  - [x] `aws-simple redeploy [options]`
- [x] Rewrite the DEV server
  - [x] `aws-simple start [options]`
- [x] Update documentation in README

### CODE REVIEW

File: src/cdk-new/create-access-log-group.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  readonly stack: Stack;
+  readonly domainName: string;
+}
+
+export function createAccessLogGroup(init: AccessLogGroupInit): ILogGroup {
+  const {stack, domainName} = init;
+
+  return new LogGroup(stack, `AccessLogGroup`, {
+    logGroupName: `/aws/apigateway/accessLogs/${domainName}}`,
+    retention: RetentionDays.TWO_WEEKS,
```
Comment (janro1) [inline src/cdk-new/create-access-log-group.ts:16]: this could be (later) come from the config

File: src/cdk-new/create-access-log-group.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+export interface AccessLogGroupInit {
+  readonly stack: Stack;
+  readonly domainName: string;
+}
+
+export function createAccessLogGroup(init: AccessLogGroupInit): ILogGroup {
+  const {stack, domainName} = init;
+
+  return new LogGroup(stack, `AccessLogGroup`, {
+    logGroupName: `/aws/apigateway/accessLogs/${domainName}}`,
```
Comment (janro1) [inline src/cdk-new/create-access-log-group.ts:15]: Is this really a domain name? or something like: `aws--simple--awesome-app-v0-33-1` ?
Ideally, you could influence the name from the config. We're currently using something like `${appName}-${appVersion}`

File: src/cdk-new/create-lambda-function.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+  }
+
+  return new LambdaFunction(stack, `LambdaFunction${getHash(functionName)}`, {
+    functionName,
+    code: Code.fromAsset(dirname(filename)),
+    handler: `${moduleName}.handler`,
+    memorySize,
+    environment,
+    timeout: Duration.seconds(Math.min(timeoutInSeconds, maxTimeoutInSeconds)),
+    runtime: Runtime.NODEJS_14_X,
```
Comment (janro1) [inline src/cdk-new/create-lambda-function.ts:54]: making this configurable would be good.

File: src/cdk-new/create-lambda-function.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    tracing: Tracing.PASS_THROUGH,
```
Comment (janro1) [inline src/cdk-new/create-lambda-function.ts:55]: same here

File: src/cdk-new/create-lambda-function.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    logRetention: RetentionDays.TWO_WEEKS,
```
Comment (janro1) [inline src/cdk-new/create-lambda-function.ts:56]: and here :)

File: src/cdk-new/create-rest-api.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+  readonly restApiName: string;
+  readonly domainName: string;
+  readonly stageOptions: StageOptions;
+}
+
+export function createRestApi(init: RestApiInit): RestApiBase {
+  const {stack, certificate, restApiName, domainName, stageOptions} = init;
+
+  const restApi = new RestApi(stack, `RestApi`, {
+    endpointTypes: [EndpointType.REGIONAL],
```
Comment (janro1) [inline src/cdk-new/create-rest-api.ts:23]: if you don't have an CDN before the stack, you might want to use edge optimized

File: src/cdk-new/create-rest-api.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    restApiName,
+    domainName: {
+      endpointType: EndpointType.REGIONAL,
+      domainName,
+      certificate,
+      securityPolicy: SecurityPolicy.TLS_1_2,
+    },
+    disableExecuteApiEndpoint: true,
+    binaryMediaTypes: [`*/*`],
```
Comment (janro1) [inline src/cdk-new/create-rest-api.ts:32]: `*/*` looks dangerous.

Comment (janro1) [reply src/cdk-new/create-rest-api.ts:32]: .. and would likely break our application

File: src/cdk-new/get-hash.ts [hunk 8]
```diff
@@ -0,0 +1,9 @@
+import {createHash} from 'crypto';
+
+export function getHash(...values: string[]): string {
+  return createHash(`sha1`)
```
Comment (janro1) [inline src/cdk-new/get-hash.ts:4]: not really important properly, but `sha1` should not be used anymore.

File: src/cdk-new/get-stage-options.ts [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+      const path = proxyName
+        ? join(publicPath, `{${proxyName}+}`, httpMethod)
+        : publicPath === `/`
+        ? `//${httpMethod}`
+        : join(publicPath, httpMethod);
+
+      return {
+        ...options,
+        [path]: {
```
Comment (janro1) [inline src/cdk-new/get-stage-options.ts:62]: don't understand the code fully, but can we do something like this with it?

```
        methodOptions: {
          '/bff/POST': {
            cachingEnabled: true,
            cacheTtl: cdk.Duration.minutes(5)
          },
          '/entry/{id}/GET': {
            cachingEnabled: false
          },
          '/pds/*/GET': {
            cachingEnabled: true,
            cacheTtl: cdk.Duration.minutes(60)
          }
        }
```

Comment (clebert) [reply src/cdk-new/get-stage-options.ts]: This is an internal function. Generally, AWS does not support `*` in paths. For that there are greedy paths like `{proxy+}`. Such kind of paths are supported. I will describe this in the documentation. It works similar to the current catchAll routes.

File: src/cdk-new/get-hash.ts (hunk 8 above)
Comment (clebert) [reply src/cdk-new/get-hash.ts]: Does not matter in non-cryptographic contexts. It is only a matter of collision avoidance here.

File: src/cdk-new/create-rest-api.ts (hunk 7 above)
Comment (clebert) [reply src/cdk-new/create-rest-api.ts]: No this is documented and has something to do with the handling of binary and non-binary files. But this takes a little longer to explain. Gladly by teams ;)

File: src/cdk-new/create-rest-api.ts (hunk 6 above)
Comment (clebert) [reply src/cdk-new/create-rest-api.ts]: Yes possibly, but that has implications. For example, the certificate must be created in US-East-1. Means 2 stacks. Until now, AWS-simple has also only supported regional endpoints. I will not change this for now. If necessary, you can put a CloudFront in front of the API gateway. I would recommend that anyway.

File: src/cdk-new/create-lambda-function.ts (hunk 5 above)
Comment (clebert) [reply src/cdk-new/create-lambda-function.ts]: OK

File: src/cdk-new/create-lambda-function.ts (hunk 4 above)
Comment (clebert) [reply src/cdk-new/create-lambda-function.ts]: No this ok. Its also the default for S3 integration. If PassThrough, Lambda will only trace the request from an upstream service if it contains a tracing header with "sampled=1". Its controlled via the TracingEnabled flag in the StackConfig.

File: src/cdk-new/create-lambda-function.ts (hunk 3 above)
Comment (clebert) [reply src/cdk-new/create-lambda-function.ts]: I think you should always use the latest Node.js version. I protect the user from an outdated version and don't let him configure it.

File: src/cdk-new/create-access-log-group.ts (hunk 2 above)
Comment (clebert) [reply src/cdk-new/create-access-log-group.ts]: Yes, according to Helge, this is the standard naming for such kind of log groups. In the future, the concept of app names and versions will no longer exist at aws-simple. The combination of HostedZoneName and possibly subdomain name is unique. Both combined correspond to the domain name used here.

File: src/cdk-new/create-access-log-group.ts (hunk 1 above)
Comment (clebert) [reply src/cdk-new/create-access-log-group.ts]: OK

File: src/cdk-new/create-lambda-function.ts (hunk 3 above)
Comment (janro1) [reply src/cdk-new/create-lambda-function.ts]: but if you update to a new version and the user is not ready or vice versa, things will break. Why not set the newest version as a default?

File: src/cdk-new/create-rest-api.ts (hunk 6 above)
Comment (janro1) [reply src/cdk-new/create-rest-api.ts]: if a user provides his own certificate ARN he could safely choose EDGE then?

File: src/cdk-new/create-lambda-function.ts (hunk 3 above)
Comment (clebert) [reply src/cdk-new/create-lambda-function.ts]: According to the library name, I will not make this configurable for now, for simplicity. If this actually leads to problems in practice, we can discuss this again via issue.

**Comment by clebert** [comment]
> Is the idea to always interact with the current stack version, or would something like this be possible?
> 
> ```
> aws-simple list | grep '1-2.*' | aws-simple tag --add some-new-tag-for-1.2xx-releases
> ```

Good idea! Done => b58002d789e6d9c53036d9e4f898c62d598391b4

**Pull request merged**